src/dom/scroll.js

```diff
@@ -0,0 +1,164 @@
+ * read or edit by the user.
+ */
+function scrollViewportToShowTarget( { target, viewportOffset = 0 } ) {
+	// Scroll the ancestors of the target to reveal it first, then focus on scrolling
+	// the viewport, when the position of the target is fixed.
+	scrollAncestorsToShowTarget( target );
+
+	const targetRect = new Rect( target );
+	const targetShiftedDownRect = targetRect.clone().moveBy( 0, viewportOffset );
+	const targetShiftedUpRect = targetRect.clone().moveBy( 0, -viewportOffset );
+	const viewportRect = Rect.getViewportRect( {
+		excludeScrollbars: true
+	} );
+
+	// Avoid the situation where the caret is still in the viewport, but totally
+	// at the edge. If it moved beyond the viewport in the next action e.g. due to enter,
+	// the scrolling would move it to the viewportOffset level and it all would look like the
+	// caret visually moved up/down.
+	//
+	// To prevent this, we're checking the rects moved by the viewportOffset to cover
+	// the upper/lower edge of the viewport.
+	const rects = [ targetShiftedUpRect, targetShiftedDownRect ];
+
+	if ( !rects.every( rect => viewportRect.contains( rect ) ) ) {
+		let { scrollX, scrollY } = global.window;
+
+		if ( isAbove( targetShiftedUpRect, viewportRect ) ) {
+			scrollY -= viewportRect.top - targetRect.top + viewportOffset;
+		} else if ( isBelow( targetShiftedDownRect, viewportRect ) ) {
+			scrollY += targetRect.bottom - viewportRect.bottom + viewportOffset;
+		}
+
+		// TODO: Web browsers scroll natively to place the target in the middle
+		// of the viewport. It's not a very popular case, though.
+		if ( isLeftOf( targetRect, viewportRect ) ) {
+			scrollX -= viewportRect.left - targetRect.left + viewportOffset;
+		} else if ( isRightOf( targetRect, viewportRect ) ) {
+			scrollX += targetRect.right - viewportRect.right + viewportOffset;
+		}
+
+		global.window.scrollTo( scrollX, scrollY );
+	}
+}
+
+/**
+ * Makes any page `HTMLElement` or `Range` (target) visible within its scrollable ancestors,
+ * e.g. if they have `overflow: scroll` CSS style.
+ *
+ * @param {HTMLElement|Range} target A target, which supposed to become visible to the user.
+ */
+function scrollAncestorsToShowTarget( target ) {
+	let parent, parentRect, targetRect;
+
+	if ( isRange( target ) ) {
+		parent = target.commonAncestorContainer;
+
+		// If a Range is attached to the Text, use the closest element ancestor.
+		if ( parent.nodeType == Node.TEXT_NODE ) {
+			parent = parent.parentNode;
+		}
+	} else {
+		parent = target.parentNode;
+	}
+
+	do {
+		if ( parent === global.document.body ) {
+			return;
+		}
+
+		targetRect = new Rect( target );
+		parentRect = new Rect( parent, {
+			excludeScrollbarsAndBorders: true
+		} );
+
+		if ( !parentRect.contains( targetRect ) ) {
+			scrollAncestorToShowTarget( { targetRect, parent, parentRect } );
+		}
+	} while ( ( parent = parent.parentNode ) );
+}
+
+export default {
```

Comment:
Why exporting this as an object? For testing purposes? Please explain it then. And in fact, it's better to expose these functions separately and then expose a test object with the two of them. So that object will be protected, while the functions will be nice and clean and exposed separately.